src/ponyfills/objects.js

```diff
@@ -0,0 +1,176 @@
+import deepmerge from 'deepmerge';
+
+export const deepMerge = deepmerge;
+export function deepClone(obj) {
+  return JSON.parse(JSON.stringify(obj));
+}
+
+/**
+ * Check if a variable is an Object. This function considers Null and Array to not be Objects.
+ *
+ * @param variable The variable that is to tested
+ * @returns {boolean} True if the variable is an object
+ */
+export function isObject(variable) {
+  return typeof variable === 'object' && variable !== null && !Array.isArray(variable);
+}
+
+/**
+ * Convert a hierarchical map into a flattened map
+ *
+ * @param {Object} map A map with hierarchical keys
+ * @returns {Object} A map with hierarchical keys flattened
+ */
+export function flatten(map, filter) {
+  function recurse(current, parent_key) {
+    let items = {};
+    Object.keys(current).filter(filter || function() { return true; }).forEach(function(k) {
+      let v = current[k];
+      let new_key = parent_key ? (parent_key + '.' + k) : k;
```

Comment:
I don't entirely recall the reason, but in the EAPI there was a need to encode the dot with the special non-visible character `\x1F` (see [PR here](https://github.com/SentientTechnologies/experiments.evolv.ai/pull/22)). I believe it was because locus path segments can include dots in them.

The approach taken in PAPI gets around this by instead representing a flattened hierarchy as an array of tuples where the first element of the tuple is an array of strings representing the path and the second element is the value of the leaf node (i.e. `type NodePath<T> = [string[], T][]`)

https://github.com/SentientTechnologies/participants.evolv.ai/blob/master/src/api/utils/flatten-object.ts

Reply:
The reason we do that in the genetics.py is because we use it to differentiate between variable keys and keys in object type variants as part of crossing and such. Differentiating was intentionally excluded as a design decision.